vsintegration/src/FSharp.Editor/Hints/HintService.fs

```diff
@@ -5,59 +5,35 @@ namespace Microsoft.VisualStudio.FSharp.Editor.Hints
-        Seq.filter (InlineParameterNameHints.isUnionCaseValidForHint symbol)
-        >> Seq.collect (InlineParameterNameHints.getHintsForUnionCase parseResults symbol)
-
-    let private getHintResolvers (sourceText: SourceText) parseResults hintKinds (symbol: FSharpSymbol) : NativeHintResolver seq =
-        let rec resolve hintKinds resolvers =
+    let private getHints sourceText parseResults hintKinds symbolUses (symbol: FSharpSymbol) =
+        let rec getHints hintKinds acc =
             match hintKinds with
-            | [] -> resolvers |> Seq.choose id
+            | [] -> acc
             | hintKind :: hintKinds ->
-                match hintKind with
-                | HintKind.TypeHint ->
-                    match symbol with
-                    | :? FSharpMemberOrFunctionOrValue as symbol -> getTypeHints parseResults symbol |> Some
-                    | _ -> None
-                | HintKind.ReturnTypeHint ->
-                    match symbol with
-                    | :? FSharpMemberOrFunctionOrValue as symbol -> getReturnTypeHints parseResults symbol |> Some
-                    | _ -> None
-                | HintKind.ParameterNameHint ->
-                    match symbol with
-                    | :? FSharpMemberOrFunctionOrValue as symbol ->
-                        getHintsForMemberOrFunctionOrValue sourceText parseResults symbol |> Some
-                    | :? FSharpUnionCase as symbol -> getHintsForUnionCase parseResults symbol |> Some
-                    | _ -> None
-                // we'll be adding other stuff gradually here
-                :: resolvers
-                |> resolve hintKinds
-
-        resolve hintKinds []
-
-    let private getHintsForSymbol (sourceText: SourceText) parseResults hintKinds (symbol: FSharpSymbol, symbolUses: FSharpSymbolUse seq) =
-        symbol
-        |> getHintResolvers sourceText parseResults hintKinds
-        |> Seq.collect (fun resolve -> resolve symbolUses)
+                match (hintKind, symbol) with
+                | HintKind.TypeHint, (:? FSharpMemberOrFunctionOrValue as symbol) ->
+                    symbolUses |> Seq.collect (InlineTypeHints(parseResults, symbol)).getHints
+                | HintKind.ReturnTypeHint, (:? FSharpMemberOrFunctionOrValue as symbol) ->
+                    symbolUses |> Seq.collect (InlineReturnTypeHints(parseResults, symbol).getHints)
+                | HintKind.ParameterNameHint, (:? FSharpMemberOrFunctionOrValue as symbol) ->
+                    symbolUses
+                    |> Seq.collect (InlineParameterNameHints(parseResults).getHintsForMemberOrFunctionOrValue sourceText symbol)
+                | HintKind.ParameterNameHint, (:? FSharpUnionCase as symbol) ->
+                    symbolUses
+                    |> Seq.collect (InlineParameterNameHints(parseResults).getHintsForUnionCase symbol)
+                | _ -> []
+                :: acc
+                |> getHints hintKinds
```

Comment:
Please, extract the match to different binding, and then use it to concat with accumulator, and then pipe into `getHints`, right now it's hard to read.